File: sdk/c/src/providers/xkx_me.c
/**
 * xkx.me 渠道实现
 *
 * 流程：
 * 1. GET 首页提取 CSRF token 和 Cookie
 * 2. POST /mailbox/create/random 创建随机邮箱
 * 3. 从 Location 或响应体提取邮箱地址
 * 4. GET /mailbox/{email}/messages 获取邮件列表
 */
#include "tempmail_internal.h"

#define XKX_BASE "https://xkx.me"

static const char *xkx_ua =
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/144.0.0.0 Safari/537.36 Edg/144.0.0.0";

/**
 * 从 HTML 中提取 CSRF token
 * 查找 csrf-token" content="..." 模式
 */
static char *xkx_extract_csrf(const char *html) {
  const char *needle = "csrf-token\" content=\"";
  char *pos = strstr(html, needle);
  if (!pos)
    return NULL;
  pos += strlen(needle);
  char *end = strchr(pos, '"');
  if (!end)
    return NULL;
  size_t len = (size_t)(end - pos);
  char *token = (char *)malloc(len + 1);
  if (!token)
    return NULL;
  memcpy(token, pos, len);
  token[len] = '\0';
  return token;
}

/**
 * 从文本中提取 xkx.me 邮箱地址
 * 查找 mailbox/xxx@xkx.me 模式
 */
static char *xkx_extract_email(const char *text) {
  const char *needle = "mailbox/";
  char *pos = strstr(text, needle);
  while (pos) {
    pos += strlen(needle);
    /* 查找 @ 和域名 */
    char *at = strchr(pos, '@');
    if (at) {
      /* 检查域名是否为 xkx.me */
      if (strncmp(at + 1, "xkx.me", 6) == 0) {
        /* 确定邮箱地址边界 */
        char *end = at + 7; /* 跳过 "xkx.me" */
        /* 确保后面是终结符 */
        if (*end == '\0' || *end == '"' || *end == '\'' || *end == '<' ||
            *end == '>' || *end == '/' || *end == ' ' || *end == '\n' ||
            *end == '\r') {
          size_t len = (size_t)(end - pos);
          char *email = (char *)malloc(len + 1);
          if (!email)
            return NULL;
          memcpy(email, pos, len);
          email[len] = '\0';
          return email;
        }
      }
    }
    /* 继续查找下一个 mailbox/ */
    pos = strstr(pos, needle);
  }
  return NULL;
}

tm_email_info_t *tm_provider_xkx_me_generate(void) {
  /* 第一步：获取首页，提取 CSRF token 和 Cookie */
  const char *get_headers[] = {"Accept: text/html", NULL};

  tm_http_response_t *resp =
      tm_http_request(TM_HTTP_GET, XKX_BASE, get_headers, NULL, 15);
  if (!resp || resp->status != 200) {
    tm_http_response_free(resp);
    return NULL;
  }

  char *csrf = xkx_extract_csrf(resp->body);
  if (!csrf) {
    tm_http_response_free(resp);
    return NULL;
  }

  /* 保存 cookie */
  char *cookies = resp->cookies ? tm_strdup(resp->cookies) : NULL;
  tm_http_response_free(resp);

  if (!cookies) {
    free(csrf);
    return NULL;
  }

  /* 第二步：POST 创建随机邮箱 */
  char body[512];
  snprintf(body, sizeof(body), "_token=%s", csrf);
  free(csrf);

  char cookie_hdr[1024];
  snprintf(cookie_hdr, sizeof(cookie_hdr), "Cookie: %s", cookies);

  char ua_hdr[256];
  snprintf(ua_hdr, sizeof(ua_hdr), "User-Agent: %s", xkx_ua);

  const char *post_headers[] = {
      cookie_hdr, "Content-Type: application/x-www-form-urlencoded", ua_hdr,
      NULL};

  tm_http_response_t *post_resp = tm_http_request(
      TM_HTTP_POST, XKX_BASE "/mailbox/create/random", post_headers, body, 15);
  if (!post_resp) {
    free(cookies);
    return NULL;
  }

  /* 从响应体中提取邮箱地址（重定向后的页面或 Location） */
  char *email = NULL;
  if (post_resp->body) {
    email = xkx_extract_email(post_resp->body);
  }

  /* 更新 cookie（如果服务端返回了新的） */
  if (post_resp->cookies && post_resp->cookies[0] != '\0') {
    free(cookies);
    cookies = tm_strdup(post_resp->cookies);
  }

  tm_http_response_free(post_resp);

  if (!email) {
    free(cookies);
    return NULL;
  }

  tm_email_info_t *info = tm_email_info_new();
  info->channel = CHANNEL_XKX_ME;
  info->email = email;
  info->token = cookies;
  return info;
}

tm_email_t *tm_provider_xkx_me_get_emails(const char *token, const char *email,
                                           int *count) {
  *count = -1;
  if (!email || !email[0])
    return NULL;

  /* 构建请求 URL */
  char url[512];
  snprintf(url, sizeof(url), XKX_BASE "/mailbox/%s/messages", email);

  /* 构建带 Cookie 的 headers */
  char cookie_hdr[1024];
  snprintf(cookie_hdr, sizeof(cookie_hdr), "Cookie: %s", token);

  char ua_hdr[256];
  snprintf(ua_hdr, sizeof(ua_hdr), "User-Agent: %s", xkx_ua);

  const char *headers[] = {cookie_hdr, "X-Requested-With: XMLHttpRequest",
                           "Accept: application/json", ua_hdr, NULL};

  tm_http_response_t *resp =
      tm_http_request(TM_HTTP_GET, url, headers, NULL, 15);
  if (!resp) {
    return NULL;
  }
  if (resp->status == 404) {
    /* 邮箱不存在或无邮件 */
    *count = 0;
    tm_http_response_free(resp);
    return NULL;
  }
  if (resp->status < 200 || resp->status >= 300) {
    tm_http_response_free(resp);
    return NULL;
  }

  /* 解析 JSON 响应 */
  cJSON *json = cJSON_Parse(resp->body);
  tm_http_response_free(resp);
  if (!json) {
    *count = 0;
    return NULL;
  }

  /* 响应可能是数组或 { "messages": [...] } 包装 */
  cJSON *arr = NULL;
  if (cJSON_IsArray(json)) {
    arr = json;
  } else {
    arr = cJSON_GetObjectItemCaseSensitive(json, "messages");
  }

  int n = (arr && cJSON_IsArray(arr)) ? cJSON_GetArraySize(arr) : 0;
  *count = n;
  if (n == 0) {
    cJSON_Delete(json);
    return NULL;
  }

  tm_email_t *emails = tm_emails_new(n);
  for (int i = 0; i < n; i++) {
    cJSON *msg = cJSON_GetArrayItem(arr, i);
    /* 构建标准化输入：优先使用 html，fallback 到 body */
    cJSON *html = cJSON_GetObjectItemCaseSensitive(msg, "html");
    if (!html || !cJSON_GetStringValue(html)) {
      cJSON *body_field = cJSON_GetObjectItemCaseSensitive(msg, "body");
      if (body_field && cJSON_GetStringValue(body_field)) {
        /* 将 body 字段值复制到 html 字段以便 normalize 处理 */
        cJSON_AddStringToObject(msg, "html", cJSON_GetStringValue(body_field));
      }
    }
    /* 添加 to 字段供 normalize 使用 */
    if (!cJSON_GetObjectItemCaseSensitive(msg, "to")) {
      cJSON_AddStringToObject(msg, "to", email);
    }
    emails[i] = tm_normalize_email(msg, email);
  }

  cJSON_Delete(json);
  return emails;
}

File: sdk/rust/src/providers/xkx_me.rs
/*!
 * XKX.me — https://xkx.me
 * 流程：GET / 获取 CSRF + cookie → POST /mailbox/create/random 创建随机邮箱
 * 收信：GET /mailbox/{email}/messages (JSON)
 */

use crate::config::{block_on, http_client};
use crate::normalize::normalize_email;
use crate::types::{Channel, Email, EmailInfo};
use regex::Regex;
use serde_json::{json, Value};

const BASE_URL: &str = "https://xkx.me";

/// 从首页 HTML 提取 CSRF token
fn extract_csrf(html: &str) -> Result<String, String> {
    let re = Regex::new(r#"csrf-token" content="([^"]+)""#).unwrap();
    let caps = re.captures(html).ok_or("xkx-me: 无法获取 CSRF token")?;
    Ok(caps[1].to_string())
}

/// 从响应头提取 Set-Cookie
fn extract_cookies(headers: &wreq::header::HeaderMap) -> String {
    headers
        .get_all("set-cookie")
        .iter()
        .filter_map(|v| v.to_str().ok())
        .map(|s| s.split(';').next().unwrap_or("").trim().to_string())
        .filter(|s| !s.is_empty())
        .collect::<Vec<_>>()
        .join("; ")
}

/// 创建 xkx.me 临时邮箱
pub fn generate_email() -> Result<EmailInfo, String> {
    block_on(async {
        let client = http_client();

        // 获取首页 CSRF + cookie
        let resp = client
            .get(BASE_URL)
            .header("Accept", "text/html")
            .header(
                "User-Agent",
                "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
            )
            .send()
            .await
            .map_err(|e| format!("xkx-me: 获取首页失败: {}", e))?;

        if !resp.status().is_success() {
            return Err(format!("xkx-me: 获取首页失败 HTTP {}", resp.status()));
        }

        let cookies = extract_cookies(resp.headers());
        let html = resp.text().await.map_err(|e| e.to_string())?;
        let csrf = extract_csrf(&html)?;

        // 创建邮箱
        let create_resp = client
            .post(format!("{}/mailbox/create/random", BASE_URL))
            .header("Content-Type", "application/x-www-form-urlencoded")
            .header("Cookie", &cookies)
            .header(
                "User-Agent",
                "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
            )
            .body(format!("_token={}", urlencoding::encode(&csrf)))
            .send()
            .await
            .map_err(|e| format!("xkx-me: 创建邮箱失败: {}", e))?;

        // 从 Location 或 body 中提取邮箱地址
        let location = create_resp
            .headers()
            .get("location")
            .and_then(|v| v.to_str().ok())
            .unwrap_or("")
            .to_string();

        let source = if location.is_empty() {
            create_resp.text().await.unwrap_or_default()
        } else {
            location
        };

        let email_re = Regex::new(r#"mailbox/([^\s/"'<>]+@xkx\.me)"#).unwrap();
        let caps = email_re
            .captures(&source)
            .ok_or("xkx-me: 无法从响应中提取邮箱地址")?;

        Ok(EmailInfo {
            channel: Channel::XkxMe,
            email: caps[1].to_string(),
            token: Some(cookies),
            expires_at: None,
            created_at: None,
        })
    })
}

/// 获取 xkx.me 邮件列表
pub fn get_emails(token: &str, email: &str) -> Result<Vec<Email>, String> {
    let address = email.trim();
    if address.is_empty() {
        return Err("xkx-me: 邮箱地址为空".into());
    }

    block_on(async {
        let client = http_client();
        let url = format!(
            "{}/mailbox/{}/messages",
            BASE_URL,
            urlencoding::encode(address)
        );

        let resp = client
            .get(&url)
            .header("Accept", "application/json")
            .header("X-Requested-With", "XMLHttpRequest")
            .header("Cookie", token)
            .header(
                "User-Agent",
                "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
            )
            .send()
            .await
            .map_err(|e| format!("xkx-me: 获取邮件失败: {}", e))?;

        if resp.status().as_u16() == 404 {
            return Ok(Vec::new());
        }
        if !resp.status().is_success() {
            return Err(format!("xkx-me: 获取邮件失败 HTTP {}", resp.status()));
        }

        let body = resp.text().await.map_err(|e| e.to_string())?;

        let messages: Vec<serde_json::Map<String, Value>> =
            serde_json::from_str(&body).unwrap_or_else(|_| {
                #[derive(serde::Deserialize)]
                struct Wrapper {
                    #[serde(default)]
                    messages: Vec<serde_json::Map<String, Value>>,
                }
                serde_json::from_str::<Wrapper>(&body)
                    .map(|w| w.messages)
                    .unwrap_or_default()
            });

        if messages.is_empty() {
            return Ok(Vec::new());
        }

        let mut out = Vec::with_capacity(messages.len());
        for msg in &messages {
            let html = msg
                .get("html")
                .or_else(|| msg.get("body"))
                .and_then(|v| v.as_str())
                .unwrap_or("");
            let raw: Value = json!({
                "id": msg.get("id"),
                "from": msg.get("from"),
                "to": address,
                "subject": msg.get("subject"),
                "date": msg.get("date"),
                "html": html,
                "text": msg.get("text").and_then(|v| v.as_str()).unwrap_or(""),
            });
            out.push(normalize_email(&raw, address));
        }
        Ok(out)
    })
}
